### DISCUSSION TITLE
ISO 8601 Date format gets http:422 Unprocessable Entity

### DISCUSSION THREAD

**Discussion by steveoriver** [fastapi/fastapi#8434] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.get("/dateParm")
def dateParm( mydate: date = Query("2022-04-06T00:00:00.000Z")) -> bool:
    return {"result":True}
```

### Description

Python FastAPI cannot process a date with the ISO8601 datetime format. I am calling the server from a Java client and am getting "422 Unprocessable Entity" This is way to strict! I know Node, Java, and Salesforce parse this correctly and truncate the time portion.

I have posted this in stackoverflow and received a workaround using union.   https://stackoverflow.com/questions/71770916/python-fastapi-iso-8601-date-format-gets-http422-unprocessable-entity

Please consider fixing this problem, and allowing the ISO format for Date type definitions.

### Operating System

Windows

### FastAPI Version

0.88.0

### Python Version

3.9

### Additional Context

StatusCode=422: Unprocessable Entity

**Comment by iudeen** [comment]
Did you try initializing a Python date object in query params as default?

**Comment by steveoriver** [comment]
All date types fail whether in a Query Param or a field within a class
object if given a full ISO string. I am just trying to get interoperability
between java and salesforce.
The example I gave was just the simplest example to reproduce the problem.

 @app.get("/dateParm")
def dateParm( mydate: date = Query("2022-04-06T00:00:00.000Z")) -> bool:

   return {"result":True}


Thanks

Steve

**Comment by iudeen** [comment]
It works fine for me. 

```python
from datetime import datetime

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/dateParm")
def date_param(mydate: datetime = Query("2022-04-06T00:00:00.000Z")):
    print(mydate)
    print(type(mydate))
    return {"result": True}
```
Gives:
```
INFO:     Started server process [15420]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)

2023-04-06 00:00:00+00:00
<class 'datetime.datetime'>
INFO:     127.0.0.1:51239 - "GET /dateParm?mydate=2023-04-06T00%3A00%3A00.000Z HTTP/1.1" 200 OK
```

The reason it fails for you is because you are trying to use `date` instead of `datetime`, in that case `2023-04-06` would work.

**Comment by steveoriver** [comment]
Your example is using a *datetime* type. It needs to be date type. The
system is being too strict, when it receives an ISO datetime for a date.

@app.get("/dateParm")
def date_param(mydate: *date* = Query("2022-04-06T00:00:00.000Z")):
    print(mydate)
    print(type(mydate))
    return {"result": True}

**Comment by iudeen** [comment]
If you pass datetime object to date variable , it is expected to fail. 

This has nothing to do with FastAPI. The validation is done by Pydantic.

**Comment by iudeen** [comment]
Try this,

```

from datetime import datetime, date

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/dateParm")
def date_param(mydate: datetime | date = Query("2022-04-06T00:00:00.000Z")):
    print(mydate)
    print(type(mydate))
    return {"result": True}

**Comment by steveoriver** [comment]
Thanks for the reply. Does Pydantic actually do the marshalling of the
data, or just the validation? If it decodes the data, do you think it's
reasonable to suggest a fix since the string is an ISO standard.
Steve

**Comment by iudeen** [comment]
Pydantic serializes and deserialize data as well. And it is rightly doing so in this case. 

In python, if you give ISO datetime string, your type should be datetime. If you want date then send only date as a string. You can always extract date from a datetime object as well.
